### PR TITLE
List exact DNS entries for a requested bucketName

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
List exact DNS entries for a requested bucketName

## Motivation and Context
Current would end up considering common prefix buckets
to be part of the same DNS service record, which leads
to Minio server wrongly forwarding the records to
incorrect IPs.

## Regression
No

## How Has This Been Tested?
Tested using following docker-compose file
```
version: '3'
services:

  minio-server-muc:
    image: custom-minio:latest
    # image: minio/minio
    ports:
      - 80:80
    volumes:
      - minio-server_muc_demo4_data_vol:/data
    entrypoint: >
        /bin/sh -c "
        curl https://raw.githubusercontent.com/nitisht/minio-compose/master/wait-for.sh -o wait-for.sh;
        chmod +x wait-for.sh;
        ./wait-for.sh etcd-server-muc:2379 -- /usr/bin/docker-entrypoint.sh minio gateway --address :80 nas /data/"
    environment:
      - "MINIO_ACCESS_KEY=minio"
      - "MINIO_SECRET_KEY=minio123"
      - "MINIO_HTTP_TRACE=/dev/stdout"
      - "MINIO_REGION=MUC"
      - "MINIO_ETCD_ENDPOINTS=http://etcd-server-muc:2379"
      - "MINIO_DOMAIN=tracecloud.test"
    networks:
      minio_network:
        ipv4_address: 172.18.0.2

  minio-server-sc:
    image: custom-minio:latest
    # image: minio/minio
    ports:
      - 90:80
    volumes:
      - minio-server_sc_demo4_data_vol:/data
    entrypoint: >
        /bin/sh -c "
        curl https://raw.githubusercontent.com/nitisht/minio-compose/master/wait-for.sh -o wait-for.sh;
        chmod +x wait-for.sh;
        ./wait-for.sh etcd-server-muc:2379 -- /usr/bin/docker-entrypoint.sh minio gateway --address :80 nas /data/"
    environment:
      - "MINIO_ACCESS_KEY=minio"
      - "MINIO_SECRET_KEY=minio123"
      - "MINIO_HTTP_TRACE=/dev/stdout"
      - "MINIO_REGION=SC"
      - "MINIO_ETCD_ENDPOINTS=http://etcd-server-muc:2379"
      - "MINIO_DOMAIN=tracecloud.test"
    networks:
      minio_network:
        ipv4_address: 172.18.0.3

  etcd-server-muc:
    image: quay.io/coreos/etcd:v3.1.10
    ports:
      - 2379:2379
      - 2380:2380
    environment:
      - "ALLOW_NONE_AUTHENTICATION=yes"
      - "ETCD_ADVERTISE_CLIENT_URLS=http://etcd-server-muc:2379"
      - "ETCD_LISTEN_CLIENT_URLS=http://172.18.0.6:2379"
    networks:
      minio_network:
        ipv4_address: 172.18.0.6

volumes:
  minio-server_muc_demo4_data_vol:
  minio-server_sc_demo4_data_vol:

networks:
 minio_network:
   driver: bridge
   ipam:
    config:
      - subnet: 172.18.0.0/24
```

```
docker-compose -f /tmp/federation.txt up
```

```
~ mc config host list fed-1
fed-1:  http://localhost  minio  minio123  s3v4  auto
~ mc mb fed-1/testbucket{1..100}
```
```
~ mc config host list fed-2
fed-2:  http://localhost:90  minio  minio123  s3v4  auto
~ mc mb fed-2/testbucket{100..200}
```

Without this change, you would see NoSuchBucket
```
~ mc ls fed-2/testbucket2
mc: <ERROR> Unable to initialize target `fed-2/testbucket2`. Bucket `testbucket2` does not exist.
```

With this change it will work fine.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#0fc98da4-7923-4777-8fed-1445050025e6) |
<a name="0fc98da4-7923-4777-8fed-1445050025e6"></a>
### 6936-0f6c5e9/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30723
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c164eaa2f9ea:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 41 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 6 seconds
(6/13) Running mc tests ... done in 36 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 42 seconds
(8/13) Running minio-go tests ... done in 27 seconds
(9/13) Running minio-java tests ... done in 2 minutes and 48 seconds
(10/13) Running minio-js tests ... FAILED in 35 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-8009b3f3-0c1b-43ac-b7d6-7a8c6924838f, events: s3:ObjectCreated:*",
  "duration": 20026,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:488:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:377:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by harshavardhana** [comment]
ping @vadmeste @nitisht

File: pkg/dns/etcd_dns.go [hunk 1]
```diff
@@ -85,7 +85,13 @@ func (c *coreDNS) list(key string) ([]SrvRecord, error) {
 		}
 		srvRecord.Key = strings.TrimPrefix(string(n.Key), key)
 		srvRecord.Key = strings.TrimSuffix(srvRecord.Key, srvRecord.Host)
-		srvRecords = append(srvRecords, srvRecord)
+		if key != msg.Path(fmt.Sprintf(".%s.", c.domainName), defaultPrefixPath) {
```
Comment (nitisht) [inline pkg/dns/etcd_dns.go:95]: Can you add comments here on why we're doing this

Comment (harshavardhana) [reply pkg/dns/etcd_dns.go:95]: Will add

**Comment by harshavardhana** [comment]
PR updated @nitisht @vadmeste

**Comment by vadmeste** [comment]
@harshavardhana currenty mint tests fail with this PR

```
(3/13) Running aws-sdk-php tests ... FAILED in 0 seconds
{
  "name": "aws-sdk-php",
  "function": "headBucket ( array $params = [] )",
  "args": [
    "aws-sdk-php-5c0fd4377045b",
    "aws-sdk-php-5c0fd437704a9"
  ],
  "duration": "26",
  "status": "FAIL",
  "error": "Error executing \"HeadBucket\" on \"http://172.18.0.3/aws-java-sdk-test-1s8rjdu\"; AWS HTTP error: Client error: `HEAD http://172.18.0.3/aws-java-sdk-test-1s8rjdu` resulted in a `404 Not Found` response NotFound (client): 404 Not Found (Request-ID: 156F50130B0C30B3) - "
}
```

**Comment by harshavardhana** [comment]
How are you running Mint tests are you using coreDNS or without it ?

**Comment by vadmeste** [comment]
> How are you running Mint tests are you using coreDNS or without it ?

With only etcd, actually I tested with docker-compose in this PR description

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by harshavardhana** [comment]
Fixed all the issues with mint @vadmeste @nitisht 

```
~ docker run --rm --net=host -e SERVER_ENDPOINT=localhost -e ACCESS_KEY=minio -e SECRET_KEY=minio123 -e ENABLE_HTTPS=0 -e MINT_MODE=full minio/mint:edge awscli aws-sdk-go aws-sdk-java aws-sdk-php aws-sdk-ruby mc minio-dotnet minio-go minio-js minio-py s3cmd security
Running with
SERVER_ENDPOINT:      localhost
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1bd627a64803:/mint/log /tmp/mint-logs'

(1/12) Running awscli tests ... done in 1 minutes and 39 seconds
(2/12) Running aws-sdk-go tests ... done in 0 seconds
(3/12) Running aws-sdk-java tests ... done in 1 seconds
(4/12) Running aws-sdk-php tests ... done in 40 seconds
(5/12) Running aws-sdk-ruby tests ... done in 2 seconds
(6/12) Running mc tests ... done in 7 seconds
(7/12) Running minio-dotnet tests ... done in 2 minutes and 27 seconds
(8/12) Running minio-go tests ... done in 8 seconds
(9/12) Running minio-js tests ... done in 22 seconds
(10/12) Running minio-py tests ... done in 42 seconds
(11/12) Running s3cmd tests ... done in 8 seconds
(12/12) Running security tests ... done in 0 seconds

All tests ran successfully
```

minio-java was selectively removed since the functional tests are not working properly yet.

**Comment by nitisht** [comment]
will test in a bit @harshavardhana

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#f828d306-cd91-449c-a482-b3a148dd341a) |
<a name="f828d306-cd91-449c-a482-b3a148dd341a"></a>
### 6936-4082673/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30679
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b981d3916847:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 14 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 7 seconds
(4/13) Running aws-sdk-php tests ... done in 43 seconds
(5/13) Running aws-sdk-ruby tests ... done in 2 seconds
(6/13) Running mc tests ... done in 16 seconds
(7/13) Running minio-dotnet tests ... FAILED in 2 minutes and 26 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=Please reduce your request)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=Please reduce your request) ---> Minio.Exceptions.MinioException: Minio API responded with message=Please reduce your request
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 363
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 483
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 229
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

**Comment by harshavardhana** [comment]
@vadmeste @nitisht it will be good to get this in for today's release. Were you able to test this already ?

**Review by nitisht — APPROVED** [review]
LGTM and tested

**Review by vadmeste — APPROVED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#f89fb458-a107-4d03-b68d-991a22ddebae) |
<a name="f89fb458-a107-4d03-b68d-991a22ddebae"></a>
### 6936-92e8976/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31661
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4367728f76c9:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 6 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 2 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 4 seconds
(6/13) Running mc tests ... done in 25 seconds
(7/13) Running minio-dotnet tests ... FAILED in 2 minutes and 39 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (Assert.Fail failed. ) ---> Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.Fail failed. 
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.HandleFail(String assertionName, String message, Object[] parameters)
   at Minio.Functional.Tests.FunctionalTest.<>c.<ListIncompleteUpload_Test2>b__85_1(Exception ex) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 2138
   at System.Reactive.AutoDetachObserver`1.OnErrorCore(Exception exception) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\AutoDetachObserver.cs:line 76
   at System.Reactive.AnonymousSafeObserver`1.OnError(Exception error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 62
   at System.Reactive.Linq.ObservableImpl.Throw`1._.Invoke() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Throw.cs:line 42
   at System.Reactive.Concurrency.Scheduler.Invoke(IScheduler scheduler, Action action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\Scheduler.Simple.cs:line 83
   at System.Reactive.Producer`2.SubscribeRaw(IObserver`1 observer, Boolean enableSafeguard) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Producer.cs:line 127
   at System.Reactive.Linq.QueryLanguage.<>c__DisplayClass330_0`1.<Create>b__0(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\QueryLanguage.Creation.cs:line 46
   at System.Reactive.AnonymousObservable`1.SubscribeCore(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousObservable.cs:line 37
   at System.Reactive.ObservableBase`1.ScheduledSubscribe(IScheduler _, AutoDetachObserver`1 autoDetachObserver) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\ObservableBase.cs:line 82
   at System.Reactive.Concurrency.CurrentThreadScheduler.Trampoline.Run(SchedulerQueue`1 queue) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\CurrentThreadScheduler.cs:line 126
   at System.Reactive.Concurrency.CurrentThreadScheduler.Schedule[TState](TState state, TimeSpan dueTime, Func`3 action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\CurrentThreadScheduler.cs:line 94
   at System.Reactive.ObservableBase`1.Subscribe(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\ObservableBase.cs:line 51
   at Minio.Functional.Tests.FunctionalTest.ListIncompleteUpload_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 2136
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 241

Executed 6 out of 13 tests successfully.
```

**Pull request merged**